### PR TITLE
Add web stream parameters

### PR DESCRIPTION
**Pull request by CumaOzavci**

### CODE REVIEW

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
Thank you! It's indeed a missing option.

However, I would rather keep the Webots format to keep consistency. That way we can simply redirect users to the official Webots documentation. What do you think?

**Comment by CumaOzavci** [comment]
I fixed consistency issues. I did not notice python version change, it must be the IDE i am using. Also i restored default values of all parameters. It should work like before now.

Keeping code simple is a great idea but i would prefer to access all Webots parameters with simple roslaunch arguments.

**Comment by lukicdarkoo** [comment]
They still would be accessible, for example:
```
roslaunch webots_ros tiago.launch mode:="port=1234,mode=x3d,monitorActivity"
```

**Comment by lukicdarkoo** [comment]
> They still would be accessible, for example:
> 
> ```
> roslaunch webots_ros tiago.launch mode:="port=1234;mode=x3d;monitorActivity"
> ```

Hello @CumaOzavci, do you agree on this? And, is it clear what we want to achieve? If you don't have time I can implement it.

**Comment by CumaOzavci** [comment]
Hi @lukicdarkoo, sorry for late reply. I was on vacation.



> They still would be accessible, for example:
> 
> ```
> roslaunch webots_ros tiago.launch mode:="port=1234;mode=x3d;monitorActivity"
> ```


I tought you were implying that **with current implementation**, they still would be accessible. Upon checking the code again, i see what you mean.

Passing all stream parameters as one string is a clever solution. I can implement it. I will commit changes this week.

By the way "mode" parameter name is already being used. I think we should name new parameter "stream" like this: 
`roslaunch webots_ros tiago.launch stream:="port=1234;mode=x3d;monitorActivity"
`

**Comment by lukicdarkoo** [comment]
@CumaOzavci

> I tought you were implying that with current implementation, they still would be accessible. Upon checking the code again, i see what you mean.

Sorry for the confusion :/

> Passing all stream parameters as one string is a clever solution. I can implement it. I will commit changes this week.

Perfect, thanks!

> By the way, "mode" parameter name is already being used. I think we should name new parameter "stream", like this:

Yes, you are right, the `stream` argument is much better, my bad.

**Comment by CumaOzavci** [comment]
@lukicdarkoo 

As you requested, i simplified web stream parameters.

Please check commits.

File: launch/webots.launch [hunk 1]
```diff
@@ -3,5 +3,6 @@
   <arg name="world" default="" doc="Path to the world to load"/>
   <arg name="mode" default="realtime" doc="Startup mode"/>
   <arg name="no_gui" default="false" doc="Start Webots with minimal GUI"/>
-  <node name="webots" pkg="webots_ros" type="webots_launcher.py" args="--world=$(arg world) --mode=$(arg mode) --no-gui=$(arg no_gui)" required="true"/>
+  <arg name="stream" default="" doc="Start Webots streaming server"/>
```
Comment (lukicdarkoo) [inline launch/webots.launch:6]:
```suggestion
  <arg name="stream" default="false" doc="Start Webots streaming server"/>
```

File: scripts/webots_launcher.py [hunk 2]
```diff
@@ -25,11 +25,17 @@
 optParser.add_option("--world", dest="world", default="", help="Path to the world to load.")
 optParser.add_option("--mode", dest="mode", default="realtime", help="Startup mode.")
 optParser.add_option("--no-gui", dest="noGui", default="false", help="Start Webots with minimal GUI.")
+optParser.add_option("--stream", dest="stream", default="", help="Start Webots streaming server.")
```
Comment (lukicdarkoo) [inline scripts/webots_launcher.py:28]:
```suggestion
optParser.add_option("--stream", dest="stream", default="false", help="Start Webots streaming server.")
```

File: scripts/webots_launcher.py [hunk 3]
```diff
@@ -25,11 +25,17 @@
...
 options, args = optParser.parse_args()
 
 if 'WEBOTS_HOME' not in os.environ:
     sys.exit('WEBOTS_HOME environment variable not defined.')
+
 command = [os.path.join(os.environ['WEBOTS_HOME'], 'webots'), '--mode=' + options.mode, options.world]
+
+if options.stream != '':
```
Comment (lukicdarkoo) [inline scripts/webots_launcher.py:36]:
```suggestion
if options.stream != 'false':
```

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
Great! 

I would change a small detail. If a user does:
```
roslaunch webots_ros tiago.launch stream:=""
```
then it should run the streaming with default parameters

**Comment by CumaOzavci** [comment]
@lukicdarkoo

> Great!
> 
> I would change a small detail. If a user does:
> 
> ```
> roslaunch webots_ros tiago.launch stream:=""
> ```
> 
> then it should run the streaming with default parameters

That's a good idea but roslaunch gives "false" value to parameters passed with "". 

Therefore i suggest this:
   1. Stream parameter should have "false" as default value. If user does not use it, streaming won't work.
   2. If user wants to stream with default values, he/she should set stream argument to "true".
   3. If user wants to stream with custom values, he/she should set stream argument to custom streaming string, like "port=4567;mode=mjpeg"

**Review by lukicdarkoo — APPROVED** [review]
Thank you!

**Comment by omichel** [comment]
@CumaOzavci: I just sent you an invite via GitHub to become a member of the Webots committers, so that you should be able to merge this PR yourself by pressing the "Merge pull request" button.

**Comment by CumaOzavci** [comment]
@omichel, thank you for adding me to Webots Committers.


As Saha Robotik, we are testing Webots for future projects. So far we are impressed with results, especially with multi robot simulations.

playground

You can expect more PRs like this, as we go further with Webots.

**Pull request merged**